### ISSUE TITLE
BTD (block tridiagonal matrices)

### ISSUE THREAD

**Issue by zerothi** [zerothi/sisl#261]
**Describe the feature**

It would be really nice to have BTD matrices in sisl, this would allow one to do RSSI much faster.

**Comment by AleksBL** [comment]
I have some code (not up to date) code that can do this on my page

**Comment by zerothi** [comment]
@AleksBL thanks for this.
I have just now uploaded a toolbox for BTD inversions. It ain't complete, but it provides details and the resources to continue work. I saw your code, does it calculate the pivoting indices for arbitrary sparse matrices, or do users have to provide pivoting indices?

**Comment by AleksBL** [comment]
Hi, it does no pivotting, but can can do all k-points and E-points in one go with numpy broadcasting when (Ev) * S  - H - Self energies have been built

**Comment by zerothi** [comment]
What do you mean *in one go*? Do you mean calculating all BTD matrices for different E and k in the same call and same class? That sounds like it would kill memory for large structures?

**Comment by AleksBL** [comment]
Yes, it does indeed do that if you dont balance how many k and E points you include in the matrix. The thought was the code could fit somewhere between the numpy ndarray and the scipy sparse class. but your implementation using sparse matrices are better when it comes to this. :)

**Comment by zerothi** [comment]
Hmm... I guess your implementation is very fast for smaller matrices. But more or less the same for large matrices where you have to them one at a time anyways?

**Comment by zerothi** [comment]
But I don't understand a great deal of your code ;)
I guess you do everything in one call, no loops for different k and E? Otherwise I wouldn't imagine there is any benefit of this?

**Comment by AleksBL** [comment]
No, that is kinda it... :)  on a periferal note, the code nests pretty well, i.e BTD matrices of BTD matrices...

**Comment by zerothi** [comment]
Is there a particular use case, other than speed-up for small systems?

**Comment by AleksBL** [comment]
Yes, In Harmonically driven systems the Hamiltonian can be written as a BTD matrix with the DFT hamiltonian on the diagonal and some couplings in between. (Its called a floquet hamiltonian, see http://homes.nano.aau.dk/tgp/master_2021.pdf  , page 398)

**Comment by zerothi** [comment]
I am not sure I follow. Is the Floquet a BTD matrix, or a set of nested BTD matrices?

**Comment by AleksBL** [comment]
Yes, the Floquet Hamiltonian is a BTD Hamiltonian, if it is driven at one frequency, It is also a nested BTD if the regular DFT Hamiltonian without any driving is BTD.

**Comment by zerothi** [comment]
I don't really follow ;)
From what I read, then one might as well do multiple BTD's matrices. There is no benefit from abstracting the nestedness?

**Comment by AleksBL** [comment]
This is going out on a bit of a tangent sorry, but im pretty sure one can calculate specific blocks  of a nested BTD ( eg. the (0,0)'th block and the (N,N)'th block) by only storing the subblocks Xt and Yt, the diagonal subblocks of the inverse of the block and 1 temporary subblock, since the action of iBTD on an arbitrary matrix can be calculated  from these and propagating the temporary element around with Xt and Yt. I think there is a RAM advantage to it,

**Comment by zerothi** [comment]
Hmm, perhaps if you wrote this in some document I could have a look. I don't really understand ;)

**Comment by AleksBL** [comment]
Hi, It is probably not doable now I have looked a bit more into it, because when one has to do eg. Yt_n = Solve( A_n - Y_n,  C_n), (A_n - Y_n) is not a BTD matrix anymore so one cant use the BTD-matrix-techniques, which i had not though about.

**Comment by zerothi** [comment]
> Hi, It is probably not doable now I have looked a bit more into it, because when one has to do eg. Yt_n = Solve( A_n - Y_n, C_n), (A_n - Y_n) is not a BTD matrix anymore so one cant use the BTD-matrix-techniques, which i had not thought about.

Ok. No, that is true, that can't be done.
